File: haskell/src/Solutions/Euler042.hs
module Solutions.Euler042 where
import Utils.Input(getInput)
import Utils.List(splitOn)
import Solutions.Euler012(triangleNumbers)
import Data.Char

tr = take 20 triangleNumbers
triangleWord w = sum (map (\c -> ord c - ord 'A' + 1) w) `elem` tr

euler042 :: IO String
euler042 = show . length . filter triangleWord . map (tail . init) . splitOn ',' <$> getInput 42

File: rust/src/solutions/euler042.rs
use crate::utils::input::get_input;
use crate::utils::numeric::sequences::triangular::Triangular;

fn word_value(word: &str) -> usize {
    word.chars().map(|c| c as usize - 'A' as usize + 1).sum()
}

fn is_triangle_word(word: &&str) -> bool {
    Triangular::new()
        .take(20)
        .collect::<Vec<usize>>()
        .contains(&word_value(*word))
}

pub fn euler042() -> String {
    get_input(42)
        .split(',')
        .map(|l| l.trim_matches('"'))
        .filter(is_triangle_word)
        .count()
        .to_string()
}
